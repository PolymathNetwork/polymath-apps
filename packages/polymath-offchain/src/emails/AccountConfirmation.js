// @flow

// TODO @monitz87: check and refactor if needed

import React from 'react';

import { EmailWrapper } from './EmailWrapper';

type Props = {|
  pin: string,
|};

export const AccountConfirmation = ({ pin }: Props) => (
  <EmailWrapper>
    <h1>Thank You for Signing Up to Polymath!</h1>
    <h2>
      To complete your sign up, please verify your email by entering the
      following PIN into your Polymath dashboard. If the correct PIN is entered,
      you will be able to proceed.
    </h2>
    <div className="main-value">
      <h2>Your Polymath Verification PIN</h2>
      <p className="value">{pin}</p>
    </div>
  </EmailWrapper>
);
