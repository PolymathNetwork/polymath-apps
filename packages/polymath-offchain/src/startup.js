import dotenv from 'dotenv';

// Initialize env variables from .env files
dotenv.config();
dotenv.config({ path: './.env.local' });
