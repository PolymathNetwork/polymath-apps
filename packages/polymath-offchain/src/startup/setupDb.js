// @flow

import mongoose from 'mongoose';
import logger from 'winston';
import P from 'bluebird';
import { MONGODB_URI } from '../constants';

mongoose.Promise = P;

mongoose.connection.once('open', () =>
  logger.info('[SETUP] Database connected.')
);
mongoose.connection.on('error', err => logger.error(err));

mongoose.connect(
  MONGODB_URI,
  { useNewUrlParser: true }
);
