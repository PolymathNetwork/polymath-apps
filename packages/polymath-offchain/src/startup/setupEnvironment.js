// @flow

import dotenv from 'dotenv';

dotenv.config();
dotenv.config({ path: './.env.local' });
