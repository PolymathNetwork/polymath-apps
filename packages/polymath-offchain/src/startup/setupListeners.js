// @flow

import { STO_MODULE_TYPE } from '../constants';
import {
  web3Client,
  getNetworkId,
  sendSTOScheduledEmail,
  sendTickerReservedEmail,
  sendTokenCreatedEmail,
} from '../utils';
import logger from 'winston';
import { User } from '../models';
import TickerRegistryArtifact from '@polymathnetwork/shared/fixtures/contracts/TickerRegistry.json';
import SecurityTokenRegistryArtifact from '@polymathnetwork/shared/fixtures/contracts/SecurityTokenRegistry.json';
import SecurityTokenArtifact from '@polymathnetwork/shared/fixtures/contracts/SecurityToken.json';
import CappedSTOArtifact from '@polymathnetwork/shared/fixtures/contracts/CappedSTO.json';

// TODO @monitz87: remake this when we rework polymath-js
// TODO @monitz87: add logs when listeners are ready

/**
  Creates a security token web3 contract object from an address

  @param {string} address contract address
 */
const getTokenContract = address => {
  return new web3Client.eth.Contract(SecurityTokenArtifact.abi, address);
};

/**
  Get details of a Capped STO
 */
const getCappedSTODetails = async (address: string) => {
  const CappedSTOContract = new web3Client.eth.Contract(
    CappedSTOArtifact.abi,
    address
  );

  try {
    const details = await CappedSTOContract.methods.getSTODetails().call();
    const fundsReceiver: string = await CappedSTOContract.methods
      .wallet()
      .call();

    /**
      STO details are returned as an object with numerical keys ranging from 0 to 7,
      we are interested in the following

      0: start time
      2: token cap for sale
      3: rate (tokens per wei / POLY)
      7: isPolyFundraise (if the fund raising is in POLY)

      Start time is a timestamp in seconds
     */
    const start = new Date(details[0] * 1000);
    const cap: number = details[2];
    const rate: number = details[3];
    const isPolyFundraise: boolean = details[7];

    return {
      start,
      cap,
      rate,
      isPolyFundraise,
      fundsReceiver,
    };
  } catch (error) {
    logger.error(error.message);
    return null;
  }
};

/**
  Add a one-time listener to a security token that triggers
  when it has been scheduled for STO

  @param contract security token web3 contract 
  @param ticker security token ticker
 */
const addSTOListener = (contract, ticker: string) => {
  contract.once(
    'LogModuleAdded',
    {
      filter: {
        _type: STO_MODULE_TYPE,
      },
    },
    async (error, result) => {
      const {
        returnValues: { _name, _module: moduleAddress },
        transactionHash,
      } = result;

      logger.info(`[EVENT] STO scheduled for "${ticker}"`);

      const moduleName: string = web3Client.utils.hexToUtf8(_name);

      /**
        Don't send an email for non CappedSTO modules
      */
      if (moduleName !== 'CappedSTO') {
        return;
      }
      /**
        Get the details of the STO
      */
      const details = await getCappedSTODetails(moduleAddress);

      if (!details) {
        return;
      }

      const { start, cap, rate, fundsReceiver, isPolyFundraise } = details;

      /**
        Get the token issuer

        TODO @monitz87: find out if the funds receiver always has the same address as the issuer
      */
      const userAddress: string = await contract.methods.owner().call();
      const user = await User.findOne({ address: userAddress });

      if (!user) {
        logger.error(`Owner not found for ${ticker}`);
        return;
      }

      const { email, name } = user;

      sendSTOScheduledEmail(
        email,
        name,
        transactionHash,
        ticker,
        cap,
        fundsReceiver,
        isPolyFundraise,
        rate,
        start
      );
    }
  );
  logger.info(`[SETUP] Listening for STO for ${ticker}`);
};

/**
  Listen for registered tickers
*/
const addTickerRegisterListener = async () => {
  const networkId = await getNetworkId();

  const TickerRegistryContract = new web3Client.eth.Contract(
    TickerRegistryArtifact.abi,
    TickerRegistryArtifact.networks[networkId].address
  );

  TickerRegistryContract.events.LogRegisterTicker({}, async (error, result) => {
    if (error) {
      logger.error(error.message);
      return;
    }

    const {
      returnValues: { _symbol: ticker, _owner: userAddress },
      transactionHash,
    } = result;

    logger.info(`[EVENT] Ticker "${ticker}" registered`);

    /**
      Get expiry limit in seconds from Ticker Registry
     */
    const expiryLimitSeconds: number = await TickerRegistryContract.methods
      .expiryLimit()
      .call();
    const expiryLimit = expiryLimitSeconds / 60 / 60 / 24;

    /**
      Get the token issuer
     */
    const user = await User.findOne({ address: userAddress });

    if (!user) {
      logger.error(`Owner not found for "${ticker}"`);
      return;
    }

    const { email, name } = user;

    sendTickerReservedEmail(email, name, transactionHash, ticker, expiryLimit);
  });
  logger.info(`[SETUP] Listening for registered tickers`);
};

/**
  Listen for newly deployed security tokens. Every time a new token gets deployed,
  we also add an STO schedule listener to it

  @param contract Security Token Registry contract
*/
const addTokenCreateListener = async contract => {
  contract.events.LogNewSecurityToken({}, async (error, result) => {
    if (error) {
      logger.error(error.message);
      return;
    }

    const {
      returnValues: { _securityTokenAddress, _ticker, _owner: userAddress },
      transactionHash,
    } = result;

    logger.info(`[EVENT] token "${_ticker}" deployed`);

    /**
        Get the token issuer
      */
    const user = await User.findOne({ address: userAddress });

    if (!user) {
      logger.error(`Owner not found for ${_ticker}`);
      return;
    }

    const { email, name } = user;

    sendTokenCreatedEmail(email, name, transactionHash, _ticker);

    const TokenContract = getTokenContract(_securityTokenAddress);

    addSTOListener(TokenContract, _ticker);
  });
  logger.info(`[SETUP] Listening for Security Token deployments`);
};

/**
  Get previously deployed security tokens and add listeners for STO scheduling

  @param contract Security Token Registry contract
*/
const addSTOListeners = async contract => {
  try {
    const previousTokenEvents = await contract.getPastEvents(
      'LogNewSecurityToken',
      {
        fromBlock: 0,
        toBlock: 'latest',
      }
    );

    for (let event of previousTokenEvents) {
      const {
        returnValues: { _securityTokenAddress, _ticker },
      } = event;

      const TokenContract = getTokenContract(_securityTokenAddress);

      addSTOListener(TokenContract, _ticker);
    }
  } catch (error) {
    logger.error(error.message);
  }
};

/**
  Asynchronously set up listeners on the blockchain to send emails
  to issuers on the following events:

  - Ticker registered
  - Security token created
  - STO scheduled
 */
const setupListeners = async () => {
  const networkId = await getNetworkId();

  await addTickerRegisterListener();

  const SecurityTokenRegistryContract = new web3Client.eth.Contract(
    SecurityTokenRegistryArtifact.abi,
    SecurityTokenRegistryArtifact.networks[networkId].address
  );

  await addTokenCreateListener(SecurityTokenRegistryContract);

  await addSTOListeners(SecurityTokenRegistryContract);
};

export default setupListeners;
