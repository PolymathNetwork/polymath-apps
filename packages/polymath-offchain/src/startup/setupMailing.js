// @flow

import logger from 'winston';
import sgMail from '@sendgrid/mail';

if (process.env.SENDGRID_API_KEY) {
  sgMail.setApiKey(process.env.SENDGRID_API_KEY);
  logger.info('[STARTUP] Mailing configured');
} else {
  logger.warn('Missing env variable SENDGRID_API_KEY.');
}
