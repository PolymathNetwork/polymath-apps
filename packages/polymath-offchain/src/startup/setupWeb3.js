// @flow

import { STO_MODULE_TYPE, WEB3_NETWORK_WS } from '../constants';
import {
  sendSTOScheduledEmail,
  sendTickerReservedEmail,
  sendTokenCreatedEmail,
} from '../utils';
import logger from 'winston';
import Web3 from 'web3';
import { User } from '../models';
import TickerRegistryArtifact from '@polymathnetwork/shared/fixtures/contracts/TickerRegistry.json';
import SecurityTokenRegistryArtifact from '@polymathnetwork/shared/fixtures/contracts/SecurityTokenRegistry.json';
import SecurityTokenArtifact from '@polymathnetwork/shared/fixtures/contracts/SecurityToken.json';
import CappedSTOArtifact from '@polymathnetwork/shared/fixtures/contracts/CappedSTO.json';

// TODO @monitz87: remake this when we rework polymath-js
let web3Client = new Web3();

/**
  Get the current network id

  @returns {number} network id
 */
const getNetworkId = async () => {
  return await web3Client.eth.net.getId();
};

/**
  Get the corresponding Security Token contract

  @param {string} address

  @returns a web3 Security Token contract
 */
const getSTContract = (address: string) => {
  return new web3Client.eth.Contract(SecurityTokenArtifact.abi, address);
};

/**
  Get the corresponding Capped STO contract

  @param {string} address

  @returns a web3 Capped STO contract
 */
const getCSTOContract = (address: string) => {
  return new web3Client.eth.Contract(CappedSTOArtifact.abi, address);
};

/**
  Get the Ticker Registry contract

  @returns a web3 Ticker Registry contract
 */
const getTRContract = async () => {
  const networkId = await getNetworkId();

  return new web3Client.eth.Contract(
    TickerRegistryArtifact.abi,
    TickerRegistryArtifact.networks[networkId].address
  );
};

/**
  Get the Security Token Registry contract

  @returns a web3 Ticker Registry contract
 */
const getSTRContract = async () => {
  const networkId = await getNetworkId();

  return new web3Client.eth.Contract(
    SecurityTokenRegistryArtifact.abi,
    SecurityTokenRegistryArtifact.networks[networkId].address
  );
};

/**
  Initializes and configures the WebsocketProvider
  for the web3, setting listeners to reconnect on error.

  NOTE @monitz87:
  This is a hack to fix a current implementation limitation of web3,
  which doesn't reconnect sockets nor re-subscribes to events when the
  socket connection is closed
 */
const newProvider = () => {
  const provider = new web3Client.providers.WebsocketProvider(WEB3_NETWORK_WS);

  /**
    Reconnect when socket connection errors or ends
   */
  provider.on('error', error => {
    logger.error(error.message, error);
    logger.info(`[SETUP] Reconnecting socket after error...`);
    connectWeb3();
  });
  provider.on('close', () => {
    logger.info(`[SETUP] Reconnecting socket after close...`);
    connectWeb3();
  });
  provider.on('end', () => {
    logger.info(`[SETUP] Reconnecting socket after end...`);
    connectWeb3();
  });

  return provider;
};

/**
  Get details of a Capped STO from the blockchain

  @param {string} address

  @returns an object with the STO details:

    start (start date),
    cap (maximum amount of tokens to sell),
    rate (how many tokens for 1 ETH/POLY),
    isPolyFundraise (is the currency POLY or ETH),
    fundsReceiver (wallet to which the funds will be transfered)
 */
const getCappedSTODetails = async (address: string) => {
  const contract = getCSTOContract(address);

  try {
    const details = await contract.methods.getSTODetails().call();
    const fundsReceiver: string = await contract.methods.wallet().call();

    /**
      STO details are returned as an object with numerical keys ranging from 0 to 7,
      we are interested in the following

      0: start time
      2: token cap for sale
      3: rate (tokens per wei / POLY)
      7: isPolyFundraise (if the fund raising is in POLY)

      Start time is a timestamp in seconds
     */
    const start = new Date(details[0] * 1000);
    const cap: number = details[2];
    const rate: number = details[3];
    const isPolyFundraise: boolean = details[7];

    return {
      start,
      cap,
      rate,
      isPolyFundraise,
      fundsReceiver,
    };
  } catch (error) {
    logger.error(error.message, error);
    return null;
  }
};

/**
  Add a listener to a security token that triggers
  when it has been scheduled for STO

  @param contract security token web3 contract 
  @param ticker security token ticker
 */
const addSTOListener = (contract, ticker: string) => {
  contract.events.LogModuleAdded(
    {
      filter: {
        _type: STO_MODULE_TYPE,
      },
    },
    async (error, result) => {
      if (error) {
        logger.error(error.message, error);
        return;
      }

      const {
        returnValues: { _name, _module: moduleAddress },
        transactionHash,
      } = result;

      logger.info(`[EVENT] STO scheduled for "${ticker}"`);

      const moduleName: string = web3Client.utils.hexToUtf8(_name);

      /**
        Don't send an email for non CappedSTO modules
      */
      if (moduleName !== 'CappedSTO') {
        return;
      }
      /**
        Get the details of the STO
      */
      const details = await getCappedSTODetails(moduleAddress);

      if (!details) {
        return;
      }

      const { start, cap, rate, fundsReceiver, isPolyFundraise } = details;

      /**
        Get the token issuer

        TODO @monitz87: find out if the funds receiver always has the same address as the issuer
      */
      const userAddress: string = await contract.methods.owner().call();
      const user = await User.findOne({ address: userAddress });

      if (!user) {
        logger.error(`Owner not found for ${ticker}`);
        return;
      }

      const { email, name } = user;

      sendSTOScheduledEmail(
        email,
        name,
        transactionHash,
        ticker,
        cap,
        fundsReceiver,
        isPolyFundraise,
        rate,
        start
      );
    }
  );

  logger.info(`[SETUP] Listening for STO for ${ticker}`);
};

/**
  Listen for registered tickers
*/
const addTickerRegisterListener = async () => {
  const contract = await getTRContract();

  contract.events.LogRegisterTicker({}, async (error, result) => {
    if (error) {
      logger.error(error.message, error);
      return;
    }

    const {
      returnValues: { _symbol: ticker, _owner: userAddress },
      transactionHash,
    } = result;

    logger.info(`[EVENT] Ticker "${ticker}" registered`);

    /**
      Get expiry limit in seconds from Ticker Registry
     */
    const expiryLimitSeconds: number = await contract.methods
      .expiryLimit()
      .call();
    const expiryLimit = expiryLimitSeconds / 60 / 60 / 24;

    /**
      Get the token issuer
     */
    const user = await User.findOne({ address: userAddress });

    if (!user) {
      logger.error(`Owner not found for "${ticker}"`);
      return;
    }

    const { email, name } = user;

    sendTickerReservedEmail(email, name, transactionHash, ticker, expiryLimit);
  });

  logger.info(`[SETUP] Listening for registered tickers`);
};

/**
  Listen for newly deployed security tokens. Every time a new token gets deployed,
  we also add an STO schedule listener to it

  @param contract Security Token Registry contract
*/
const addTokenCreateListener = async () => {
  const contract = await getSTRContract();

  contract.events.LogNewSecurityToken({}, async (error, result) => {
    if (error) {
      logger.error(error.message, error);
      return;
    }

    const {
      returnValues: { _securityTokenAddress, _ticker, _owner: userAddress },
      transactionHash,
    } = result;

    logger.info(`[EVENT] Token "${_ticker}" deployed`);

    /**
        Get the token issuer
      */
    const user = await User.findOne({ address: userAddress });

    if (!user) {
      logger.error(`Owner not found for ${_ticker}`);
      return;
    }

    const { email, name } = user;

    sendTokenCreatedEmail(email, name, transactionHash, _ticker);

    const tokenContract = getSTContract(_securityTokenAddress);

    addSTOListener(tokenContract, _ticker);
  });

  logger.info(`[SETUP] Listening for Security Token deployments`);
};

/**
  Get previously deployed security tokens and add listeners for STO scheduling

  @param contract Security Token Registry contract
*/
const addSTOListeners = async () => {
  const contract = await getSTRContract();
  try {
    const previousTokenEvents = await contract.getPastEvents(
      'LogNewSecurityToken',
      {
        fromBlock: 0,
        toBlock: 'latest',
      }
    );

    for (let event of previousTokenEvents) {
      const {
        returnValues: { _securityTokenAddress, _ticker },
      } = event;

      const TokenContract = getSTContract(_securityTokenAddress);

      addSTOListener(TokenContract, _ticker);
    }
  } catch (error) {
    logger.error(error.message, error);
  }
};

/**
  Asynchronously set up listeners on the blockchain to send emails
  to issuers on the following events:

  - Ticker registered
  - Security token created
  - STO scheduled
 */
const setupListeners = async () => {
  await addTickerRegisterListener();

  await addTokenCreateListener();

  await addSTOListeners();
};

let heartbeatIntervalId;

/**
  Ping the socket. If there is something wrong with the conection,
  we kill the heartbeat and reset the web3 client and all the listeners
 */
const keepAlive = async () => {
  try {
    const isListening = await web3Client.eth.net.isListening();
    if (!isListening) {
      throw new Error('Socket not listening to peers');
    }
  } catch (error) {
    logger.error(error.message, error);
    clearInterval(heartbeatIntervalId);
    // TODO @monitz87: kill the socket instead of calling connectWeb3
    connectWeb3();
  }
};

/**
  Ping socket every 5 seconds to keep it alive
  */
const simulateHeartbeat = () => {
  const heartbeatIntervalId = setInterval(keepAlive, 1000);
};

/**
  Connects the web3 client to a new provider and starts all the event listeners
 */
const connectWeb3 = async () => {
  web3Client = new Web3(newProvider());

  simulateHeartbeat();

  try {
    await setupListeners();
  } catch (error) {
    logger.error(error.message, error);
  }
};

export default connectWeb3;
