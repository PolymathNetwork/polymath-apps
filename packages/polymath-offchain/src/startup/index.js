// @flow
/* Setup scripts */
import './setupEnvironment';
import { NETWORKS } from '../constants';
import './initializeLogger';

import logger from 'winston';

import './setupDb';
import './setupMailing';
import connectWeb3 from './setupWeb3';

(async () => {
  const { '15': local, '42': kovan, '1': mainnet } = NETWORKS;
  if (local) {
    await connectWeb3('15');
  }

  if (kovan) {
    await connectWeb3('42');
  }

  if (mainnet) {
    await connectWeb3('1');
  }
})().catch(err => logger.error(err.message, err));
