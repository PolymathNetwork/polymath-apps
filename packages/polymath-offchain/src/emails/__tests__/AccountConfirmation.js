import '../../startup/setupEnvironment';

import React from 'react';
import renderer from 'react-test-renderer';
import { AccountConfirmation } from '../AccountConfirmation';

describe('Component: AccountConfirmation', () => {
  test('renders without crashing', () => {
    const component = renderer.create(
      <AccountConfirmation pin={'202bb23326ce5624'} />
    );
    let tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
