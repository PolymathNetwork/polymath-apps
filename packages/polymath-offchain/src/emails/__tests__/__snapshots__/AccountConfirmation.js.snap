// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: AccountConfirmation renders without crashing 1`] = `
<html
  lang="en"
>
  <head>
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": "
  body {
    font: small/1.5 -apple-system, BlinkMacSystemFont, Arial, Helvetica, sans-serif;
    color: #152935;
    font-size: 16px;
  }
  
  a {
    color: #3D70B2 !important;
    text-decoration: none;
  }
  
  a:hover {
    text-decoration: underline;
  }
  
  ul,
  ol {
    margin-top: 5px;
  }
  
  ul li,
  ol li {
    font-size: 18px;
    margin-bottom: 7px;
  }
  
  ol li {
    padding-left: 5px;
  }
  
  h1 {
    font-size: 36px;
    line-height: 43px;
    font-weight: 600;
    color: #252E6A;
    margin-top: 0;
    margin-bottom: 11px;
  }
  
  h2 {
    font-size: 20px;
    line-height: 28px;
    color: #152935;
    font-weight: normal;
  }
  
  h3 {
    font-size: 20px;
    font-weight: 600;
    margin-bottom: 12px;
    color: #152935 !important;
  }
  
  h4 {
    color: #3D70B2;
    font-size: 14px;
    line-height: 18px;
    font-weight: 600;
    margin-top: 0;
    margin-bottom: 0;
  }
  
  .wrapper {
    width: 768px;
    margin: 0 auto;
  }
  
  .top-bar {
    height: 48px;
    border-bottom: 1px solid #DFE3E6;
    padding-left: 72px;
    margin-bottom: 24px;
  }
  
  .top-bar img {
    margin-top: 11px;
    width: 188px;
    height: 29px;
  }
  
  .content {
    padding: 3px 72px 25px;
  }
  
  .main-value {
    width: 624px;
    height: 160px;
    background-color: #EBF0F7;
    border-radius: 4px;
    margin-bottom: 32px;
    margin-top: 28px;
  }
  
  .main-value h2 {
    line-height: 18px;
    text-align: center;
    padding-top: 19px;
    font-weight: 600;
  }
  
  .main-value .value {
    width: 560px;
    height: 64px;
    line-height: 64px;
    text-align: center;
    font-size: 20px;
    background-color: white;
    margin: 0 auto;
    vertical-align: middle;
  }
  
  .icon-text {
    margin-bottom: 30px;
  }
  
  .icon-text .icon {
    width: 48px;
    height: 100%;
    float: left;
  }
  
  .icon-text .icon.question {
    margin-top: 5px;
  }
  
  .icon-text .icon img {
    width: 32px;
    height: 32px;
    margin-top: 4px;
  }
  
  .icon-text h2 {
    margin: 0;
    line-height: 25px;
  }
  
  .icon-text p {
    margin-top: 3px;
  }
  
  .icon-text.tx-hash {
    height: 57px;
    margin-top: 20px;
  }
  
  .text {
    margin-bottom: 26px;
    color: #5A6872;
  }
  
  .text p {
    font-size: 18px;
  }
  
  .value {
    margin-top: 13px;
  }
  
  .value p {
    margin-top: 8px;
  }
  
  .value strong {
    color: #152935 !important;
  }
  
  .footer {
    height: 48px;
    border-top: 1px solid #DFE3E6;
    padding-left: 72px;
    padding-right: 72px;
    font-size: 14px;
    color: #5A6872;
    padding-top: 14px;
  }
  
  .footer .left,
  .footer .right {
    float: left;
    width: 50%;
  }
  
  .footer .right {
    text-align: right;
  }
  
  .footer a {
    color: #5A6872 !important;
    margin-left: 40px;
  }
  
  .tx {
    float: left;
    width: 300px;
    height: 25px;
    border-radius: 4px;
    background-color: #EBF0F7;
    font-size: 12px;
    line-height: 25px;
    padding: 0 16px;
    margin-top: 6px;
    text-align: center;
  }
  
  .tx a {
    text-decoration: underline;
  }
  
  .sincere {
    font-weight: 600;
    color: #252E6A;
    line-height: 31px;
  }
",
        }
      }
    />
  </head>
  <body>
    <div
      className="wrapper"
    >
      <div
        className="top-bar"
      >
        <img
          alt="Icon"
          src="http://localhost:3001/img/logo.png"
        />
      </div>
      <div
        className="content"
      >
        <h1>
          Thank You for Signing Up to Polymath!
        </h1>
        <h2>
          To complete your sign up, please verify your email by entering the following PIN into your Polymath dashboard. If the correct PIN is entered, you will be able to proceed.
        </h2>
        <div
          className="main-value"
        >
          <h2>
            Your Polymath Verification PIN
          </h2>
          <p
            className="value"
          >
            202bb23326ce5624
          </p>
        </div>
        <div
          className="icon-text"
          style={
            Object {
              "height": "52px",
            }
          }
        >
          <div
            className="icon question"
          >
            <img
              alt="Icon"
              src="http://localhost:3001/img/question.png"
            />
          </div>
          <h2>
            If you have any questions, please reach out to
            <br />
            <a
              href="mailto:tokenstudio@polymath.network"
            >
              tokenstudio@polymath.network
            </a>
          </h2>
        </div>
        <h2
          className="sincere"
        >
          Best,
          <br />
          Polymath Support
        </h2>
      </div>
      <div
        className="footer"
      >
        <div
          className="left"
        >
          Â© 2018 Polymath
        </div>
        <div
          className="right"
        >
          <a
            href="https://polymath.network/termsofservice.html"
          >
            Terms of service
          </a>
          <a
            href="https://polymath.network/privacypolicy.html"
          >
            Privacy policy
          </a>
        </div>
      </div>
    </div>
  </body>
</html>
`;
